### PR TITLE
feat(bundler): Use centralized docker execution

### PR DESCRIPTION
**Pull request by zharinov**
ref #2863

### CODE REVIEW

File: lib/manager/bundler/artifacts.ts [hunk 1]
```diff
@@ -83,43 +90,25 @@ export async function updateArtifacts(
...
-        cmd += `--user=${config.dockerUser} `;
-      }
-      const volumes = [config.localDir];
-      cmd += volumes.map(v => `-v "${v}":"${v}" `).join('');
-      cmd += `-w "${cwd}" `;
-      cmd += `renovate/ruby:${tag} bash -l -c "ruby --version && `;
-      cmd += 'gem install bundler' + bundlerVersion + ' --no-document';
-      cmd += ' && bundle';
+      const bundlerVersion = bundler ? ` -v ${bundler}` : '';
+      const cmdWrap = `bash -l -c "ruby --version && gem install bundler${bundlerVersion} --no-document && {{ cmd }}"`;
```
Comment (ViceIce) [inline lib/manager/bundler/artifacts.ts]: Use pre /post commands, as @rarkins suggested in #5061

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by zharinov** [comment]
Ready for review

File: test/manager/bundler/artifacts.spec.ts [hunk 2]
```diff
@@ -32,16 +34,21 @@ describe('bundler.updateArtifacts()', () => {
...
+      ...config,
+      dockerUser: 'foobar',
+    });
   });
   it('returns null by default', async () => {
     expect(
       await updateArtifacts({
         packageFileName: '',
-        updatedDeps: [],
+        updatedDeps: ['foo', 'bar'],
```
Comment (ViceIce) [inline test/manager/bundler/artifacts.spec.ts]: Do we have an test for no artifacts? Then we should now call `bundle lock` i think

**Review by viceice — DISMISSED** [review]

File: test/manager/bundler/artifacts.spec.ts (hunk 2 above)
Comment (zharinov) [reply test/manager/bundler/artifacts.spec.ts]: Are empty arrays possible?

Comment (ViceIce) [reply test/manager/bundler/artifacts.spec.ts]: 🤔 @rarkins do you have an answer?

**Comment by rarkins** [comment]
Let’s pause this one for now. Will require some more refactoring of the exec until function

File: lib/manager/bundler/artifacts.ts [hunk 3]
```diff
@@ -18,12 +17,63 @@ import {
 } from '../../constants/error-messages';
 import { BinarySource } from '../../util/exec/common';
 
-export async function updateArtifacts({
-  packageFileName,
-  updatedDeps,
-  newPackageFileContent,
-  config,
-}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+async function getDockerTag(updateArtifact: UpdateArtifact): Promise<string> {
```
Comment (zharinov) [inline lib/manager/bundler/artifacts.ts]:
```typescript
interface DockerOptions {
  ...
  tag?: string | (UpdateArtifact) => Promise<string>;
  ...
}
```

?

Comment (zharinov) [reply lib/manager/bundler/artifacts.ts]: But we can't provide `UpdateArtifact` argument from `dockerCmd(...)` call, so should be `(void) => Promise<string>` actually

Comment (rarkins) [reply lib/manager/bundler/artifacts.ts]: Let's calculate the tag within this file and then pass it through to util. We don't need to do the tag lookup in util yet.

File: lib/manager/bundler/artifacts.ts [hunk 4]
```diff
@@ -36,92 +86,39 @@ export async function updateArtifacts({
...
+    const { bundler } = compatibility;
+    const bundlerVersion = bundler ? ` -v ${bundler}` : '';
+    const preCommands = [
+      'ruby --version',
+      `gem install bundler${bundlerVersion} --no-document`,
+    ];
+
+    const execOptions: ExecOptions = {
+      docker: {
+        image: 'renovate/ruby',
```
Comment (rarkins) [inline lib/manager/bundler/artifacts.ts]: We should pass the calculated image tag in this line

File: test/manager/bundler/artifacts.spec.ts (hunk 2 above)
Comment (rarkins) [reply test/manager/bundler/artifacts.spec.ts]: They're not meant to be empty

**Pull request merged**